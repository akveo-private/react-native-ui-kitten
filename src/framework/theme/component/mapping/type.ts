// TODO(mapping/type): declare Config/Token type

export type ThemeMappingConfigType = any;

export interface ThemeMappingType {
  parameters: string[];
  variants: any;
}

export interface VariantType {
  state: any;
}

export type TokenType = any;
