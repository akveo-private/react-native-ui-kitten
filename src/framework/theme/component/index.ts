export { ThemeProvider } from './themeProvider.component';
export { withTheme } from './themeConsumer.component';
export { withThemedStyles } from './themeConsumerStyled.component';
export { ThemeType } from './type';
