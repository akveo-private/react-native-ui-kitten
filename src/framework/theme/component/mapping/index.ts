export * from './type';
export {
  MappingProvider as ThemeMappingProvider,
  Props as ThemeMappingProviderProps,
} from './mappingProvider.component';
