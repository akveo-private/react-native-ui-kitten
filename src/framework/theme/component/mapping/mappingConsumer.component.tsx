import React, { ComponentType } from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';
import {
  getComponentThemeMapping,
  forwardProps,
} from '../../service';

export interface Props<P> extends React.ClassAttributes<P> {
  mapping: ThemeMappingType;
}

export function withMapping<P extends Props<P>>(Component: ComponentType<P>) {
  type TExcept = Exclude<keyof P, keyof Props<P>>;
  type ForwardedProps = Pick<P, TExcept>;

  class Wrapper extends React.Component<ForwardedProps> {
    wrappedComponentRef = undefined;
    getWrappedInstance = undefined;

    setWrappedComponentRef = (ref) => {
      this.wrappedComponentRef = ref;
    };

    getComponentName = () => Component.displayName || Component.name;

    renderWrappedComponent = (mapping: ThemeMappingType[]) => (
      <Component
        ref={this.setWrappedComponentRef}
        mapping={getComponentThemeMapping(this.getComponentName(), mapping)}
        {...this.props}
      />
    );

    render() {
      return (
        <MappingContext.Consumer>
          {this.renderWrappedComponent}
        </MappingContext.Consumer>
      );
    }
  }

  const Result = Wrapper;
  Result.prototype.getWrappedInstance = function getWrappedInstance() {
    const hasWrappedInstance = this.wrappedComponentRef && this.wrappedComponentRef.getWrappedInstance;
    return hasWrappedInstance ? this.wrappedComponentRef.getWrappedInstance() : this.wrappedComponentRef;
  };

  return forwardProps(Component, Result);
}
