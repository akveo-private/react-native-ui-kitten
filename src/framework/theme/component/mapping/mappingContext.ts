import React from 'react';
import { ThemeMappingType } from './type';

const defaultValue: ThemeMappingType = {};
const MappingContext = React.createContext(defaultValue);

export default MappingContext;
