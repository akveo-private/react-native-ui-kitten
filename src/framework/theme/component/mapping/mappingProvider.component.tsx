import React from 'react';
import MappingContext from './mappingContext';
import { ThemeMappingType } from './type';

export interface Props {
  mapping: ThemeMappingType;
  children: JSX.Element | React.ReactNode;
}

export class MappingProvider extends React.PureComponent<Props> {

  render() {
    return (
      <MappingContext.Provider
        value={this.props.mapping}>
        {this.props.children}
      </MappingContext.Provider>
    );
  }
}
