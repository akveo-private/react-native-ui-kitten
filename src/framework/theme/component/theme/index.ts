export * from './type';
export {
  ThemeProvider,
  Props as ThemeProviderProps,
} from './themeProvider.component';
export {
  withStyles,
  Props as ThemedComponentProps,
} from './themeConsumer.component';
