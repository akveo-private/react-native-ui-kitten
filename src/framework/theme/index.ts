export {
  StyleProvider,
  styled,
  ThemeProvider,
  withStyles,

  StyleProviderProps,
  StyledComponentProps,
  ThemeProviderProps,
  ThemedComponentProps,

  ThemeType,
  ThemeMappingType,
  ComponentMappingType,
  TokenType,
  ThemedStyleType,
  StyleSheetType,
  StyleType,
} from './component';

export {
  APPEARANCE_DEFAULT,
} from './service';

