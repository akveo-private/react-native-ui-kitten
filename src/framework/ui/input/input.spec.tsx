import React from 'react';
import {
  Image,
  ImageProps,
  ImageSourcePropType,
  TextInput,
} from 'react-native';
import {
  render,
  fireEvent,
  shallow,
  RenderAPI,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import {
  styled,
  ApplicationProvider,
  ApplicationProviderProps,
  StyleType,
} from '@kitten/theme';
import {
  Input as InputComponent,
  Props as InputProps,
} from './input.component';
import { default as mapping } from '../common/mapping.json';
import { default as theme } from '../common/theme.json';

const Input = styled<InputProps>(InputComponent);

const Mock = (props?: InputProps): React.ReactElement<ApplicationProviderProps> => {
  return (
    <ApplicationProvider
      mapping={mapping}
      theme={theme}>
      <Input {...props}/>
    </ApplicationProvider>
  );
};

const renderComponent = (props?: InputProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@input: matches snapshot', () => {

  describe('* interaction', () => {

    it('* stateless', () => {
      const component: RenderAPI = renderComponent();

      const { output } = shallow(component.getByType(InputComponent));

      expect(output).toMatchSnapshot();
    });

  });

  describe('* appearance', () => {

    const iconSource: ImageSourcePropType = { uri: 'https://akveo.github.io/eva-icons/fill/png/128/star.png' };

    it('* icon', () => {
      const icon = (style: StyleType): React.ReactElement<ImageProps> => {
        return (
          <Image
            style={style}
            source={iconSource}
          />
        );
      };

      const component: RenderAPI = renderComponent({ icon });

      const { output } = shallow(component.getByType(InputComponent));

      expect(output).toMatchSnapshot();
    });

    it('* label + caption', () => {
      const captionIcon = (style: StyleType): React.ReactElement<ImageProps> => {
        return (
          <Image
            style={style}
            source={iconSource}
          />
        );
      };
      const label: string = 'Label';
      const caption: string = 'Caption Text';

      const component: RenderAPI = renderComponent({
        label,
        caption,
        captionIcon,
      });

      const { output } = shallow(component.getByType(InputComponent));

      expect(output).toMatchSnapshot();
    });

  });

});

interface InputListenerState {
  text: string;
}

class InputListener extends React.Component<InputProps, InputListenerState> {

  public state: InputListenerState = {
    text: '',
  };

  private onChangeText = (text: string) => {
    this.setState({ text });
  };

  public render(): React.ReactNode {
    return (
      <Mock
        {...this.props}
        value={this.state.text}
        onChangeText={this.onChangeText}
      />
    );
  }
}

describe('@input: component checks', () => {

  it('* emits onChangeText', () => {
    const onChangeText = jest.fn();

    const component: RenderAPI = renderComponent({ onChangeText });

    fireEvent.changeText(component.getByType(TextInput), 'it works!');

    expect(onChangeText).toBeCalledWith('it works!');
  });

  it('* emits onFocus', () => {
    const onFocus = jest.fn();

    const component: RenderAPI = renderComponent({ onFocus });

    fireEvent(component.getByType(TextInput), 'focus');

    expect(onFocus).toBeCalled();
  });

  it('* emits onEndEditing', () => {
    const onEndEditing = jest.fn();

    const component: RenderAPI = renderComponent({ onEndEditing });

    fireEvent(component.getByType(TextInput), 'endEditing');

    expect(onEndEditing).toBeCalled();
  });

  it('* changes text', () => {
    const component: RenderAPI = render(
      <InputListener/>,
    );

    const input: ReactTestInstance = component.getByType(TextInput);

    fireEvent.changeText(input, 'it works!');

    const updatedInput: ReactTestInstance = component.getByType(TextInput);

    expect(updatedInput.props.value).toEqual('it works!');
  });

});
