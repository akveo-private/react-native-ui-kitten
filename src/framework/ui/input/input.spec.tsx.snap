// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@input: matches snapshot * appearance * icon 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#f7f8fa",
        "borderColor": "transparent",
        "borderRadius": 6,
        "borderWidth": 2,
        "minHeight": 48,
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    appearance="default"
    dispatch={[Function]}
    editable={true}
    icon={[Function]}
    onEndEditing={[Function]}
    onFocus={[Function]}
    status="primary"
    style={
      Array [
        Object {
          "color": "#a6aebd",
          "marginHorizontal": 6,
        },
        Object {
          "color": undefined,
          "fontFamily": undefined,
          "fontSize": undefined,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "letterSpacing": undefined,
          "textAlign": undefined,
        },
        Object {
          "flex": 1,
        },
      ]
    }
    theme={
      Object {
        "blue-dark": "#2541CC",
        "blue-pitch-dark": "#0D1C2E",
        "blue-primary": "#3366FF",
        "gray-100": "#f7f8fa",
        "gray-200": "#edf0f5",
        "gray-300": "#c8cedb",
        "gray-400": "#a6aebd",
        "gray-dark": "#8992A3",
        "gray-highlight": "#EDF0F5",
        "gray-light": "#DDE1EB",
        "gray-primary": "#A6AEBD",
        "pink-primary": "#FF3D71",
        "text-primary": "#000000",
        "text-primary-inverse": "#ffffff",
      }
    }
    underlineColorAndroid="transparent"
  />
  <Component
    source={
      Object {
        "uri": "https://akveo.github.io/eva-icons/fill/png/128/star.png",
      }
    }
    style={
      Object {
        "height": 24,
        "marginHorizontal": 6,
        "tintColor": "#a6aebd",
        "width": 24,
      }
    }
  />
</View>
`;

exports[`@input: matches snapshot * interaction * stateless 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#f7f8fa",
        "borderColor": "transparent",
        "borderRadius": 6,
        "borderWidth": 2,
        "minHeight": 48,
        "paddingHorizontal": 6,
        "paddingVertical": 12,
      },
      Object {},
      Object {
        "flex": 1,
        "flexDirection": "row",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    appearance="default"
    dispatch={[Function]}
    editable={true}
    onEndEditing={[Function]}
    onFocus={[Function]}
    status="primary"
    style={
      Array [
        Object {
          "color": "#a6aebd",
          "marginHorizontal": 6,
        },
        Object {
          "color": undefined,
          "fontFamily": undefined,
          "fontSize": undefined,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "letterSpacing": undefined,
          "textAlign": undefined,
        },
        Object {
          "flex": 1,
        },
      ]
    }
    theme={
      Object {
        "blue-dark": "#2541CC",
        "blue-pitch-dark": "#0D1C2E",
        "blue-primary": "#3366FF",
        "gray-100": "#f7f8fa",
        "gray-200": "#edf0f5",
        "gray-300": "#c8cedb",
        "gray-400": "#a6aebd",
        "gray-dark": "#8992A3",
        "gray-highlight": "#EDF0F5",
        "gray-light": "#DDE1EB",
        "gray-primary": "#A6AEBD",
        "pink-primary": "#FF3D71",
        "text-primary": "#000000",
        "text-primary-inverse": "#ffffff",
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
